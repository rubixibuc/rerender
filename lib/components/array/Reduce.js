import React from "react";
import PropTypes from "prop-types";

const render = ({ children, initial, items, reducer }) =>
  (items || null) && children(items.reduce(reducer, initial));

render.propTypes = {
  children: PropTypes.func.isRequired,
  initial: PropTypes.any,
  items: PropTypes.array,
  reducer: PropTypes.func.isRequired
};

render.defaultProps = {
  initial: {}
};

export default render;
