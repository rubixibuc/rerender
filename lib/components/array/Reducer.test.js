import React from "react";
import Reduce from "./Reduce";
import renderer from "react-test-renderer";

test("should render nothing when items is undefined", () => {
  const component = renderer.create(
    <Reduce reducer={i => i} items={undefined}>
      {items => <div>{items}</div>}
    </Reduce>
  );
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

test("should render nothing when items is null", () => {
  const component = renderer.create(
    <Reduce reducer={i => i} items={null}>
      {items => <div>{items}</div>}
    </Reduce>
  );
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

test("should render nothing when items is empty", () => {
  const component = renderer.create(
    <Reduce reducer={i => i} items={[]}>
      {items => <div>{JSON.stringify(items)}</div>}
    </Reduce>
  );
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

test("should render reduced items is when items is not empty", () => {
  const component = renderer.create(
    <Reduce
      reducer={(accumulator, value) => {
        accumulator[value] = value;
        return accumulator;
      }}
      items={[1, 2]}
    >
      {items => <div>{JSON.stringify(items)}</div>}
    </Reduce>
  );
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});
