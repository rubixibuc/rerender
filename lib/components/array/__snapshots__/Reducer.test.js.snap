// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render nothing when items is empty 1`] = `
<div>
  {}
</div>
`;

exports[`should render nothing when items is null 1`] = `null`;

exports[`should render nothing when items is undefined 1`] = `null`;

exports[`should render reduced items is when items is not empty 1`] = `
<div>
  {"1":1,"2":2}
</div>
`;
